pynest/nest/lib/hl_api_spatial.py

```diff
@@ -728,6 +731,81 @@ def GetTargetNodes(sources, tgt_layer, syn_model=None):
     return tuple(src_tgt_map[snode_id] for snode_id in sources.tolist())
 
 
+def GetSourceNodes(src_layer, targets, syn_model=None):
+    """
+    Obtain sources of `targets` in given `source` population.
+
+    For each neuron in `targets`, this function finds all target elements
+    in `src_layer`. If `syn_model` is not given (default), all sources are
+    returned, otherwise only targets of specific type.
+
+    Parameters
+    ----------
+    src_layer : NodeCollection
+        NodeCollection with node IDs of `src_layer`
+    targets : NodeCollection
+        NodeCollection with node IDs of `targets`
+    syn_model : [None | str], optional, default: None
+        Return only target positions for a given synapse model.
+
+    Returns
+    -------
+    tuple of NodeCollection:
+        Tuple of `NodeCollections` of source neurons fulfilling the given criteria, one `NodeCollection` per
+        target node ID in `target`.
+
+    See also
+    --------
+    GetSourcePositions: Obtain positions of sources in a given source layer connected to given target.
+    GetConnections: Return connection identifiers between
+        sources and targets
+
+    Notes
+    -----
+    * For distributed simulations, this function only returns source on the
+      local MPI process.
+
+    Example
+    -------
+        ::
+
+            import nest
+
+            # create a spatial population
+            s_nodes = nest.Create('iaf_psc_alpha', positions=nest.spatial.grid(shape=[11, 11], extent=[11., 11.]))
+
+            # connectivity specifications with a mask
+            conndict = {'rule': 'pairwise_bernoulli', 'p': 1.,
+                        'mask': {'rectangular': {'lower_left' : [-2.0, -1.0],
+                                                 'upper_right': [2.0, 1.0]}}}
+
+            # connect population s_nodes with itself according to the given
+            # specifications
+            nest.Connect(s_nodes, s_nodes, conndict)
+
+            # get the node IDs of the targets of a source neuron
+            nest.GetSourceNodes(s_nodes, s_nodes[4])
+    """
+    if not isinstance(src_layer, NodeCollection):
+        raise TypeError("src_layer must be a NodeCollection")
+
+    if not isinstance(targets, NodeCollection):
+        raise TypeError("targets must be a NodeCollection.")
+
+    conns = GetConnections(src_layer, targets, synapse_model=syn_model)
+
+    # Re-organize conns into one list per target, containing only source node IDs.
+    tgt_src_map = dict((tnode_id, []) for tnode_id in targets.tolist())
+    for src, tgt in zip(conns.sources(), conns.targets()):
+        tgt_src_map[tgt].append(src)
+
+    for tgt in tgt_src_map.keys():
+        tgt_src_map[tgt] = NodeCollection(list(np.unique(tgt_src_map[tgt])))
```

Comment:
How about inserting the sources into sets instead of lists, since you then get the uniqueness automatically?

Reply:
In principal I agree, this is a nice idea and the resulting code would look more pythonic.
Hence I tried it, but it does not work.
The point where it seems to break is at `sli_push`, this cannot work with Python sets and throws
`PyNESTError: unknown Python type: <class 'set'>`.

I thus think that this suggestion goes beyond the scope of this PR.
What do you think?

Reply:
Then just leave it as it is, I hadn't thought about that complication.